### PR TITLE
Store refactor: make it easier to test stores

### PR DESCRIPTION
**Pull request by kegsay**
This is the first PR in a series of refactors with the aim of making stores more consistent with each other and easier to test. 

Main changes:
 - Expand the already existing `untilDispatch` to allow for more thorough checks and a built-in default timeout, and add `untilEmission` for the `EventEmitter` equivalent. This makes it easier and more robust to test stores for well-defined events rather than assuming implementation details by calling `flushPromises()` which doesn't work for re-dispatch calls as dispatches use `setTimeout`.
 - Convert RVS from a Flux store to a general EventEmitter. This makes it more consistent with other stores which similarly use UPDATE_EVENT, and allows us to remove boilerplate room listener code whilst preserving functionality.
 - Add constructor dependency injection to RVS for the Dispatcher only for now. This makes tests more robust because it ensures that a clean, unpolluted dispatcher is used for every test. Previously, this was not guaranteed, so if a test caused something to dispatch asyncly (e.g a timer firing) it could be consumed in a different test. This will be expanded in the future to include the remaining dependencies: PosthogAnalytics, SpaceStore, SlidingSyncManager, MatrixClientPeg and SettingsStore.
 - Added a battery of RoomViewStore tests for JoinRoomError and ViewRoomError, `auto_join` and others to try it out.
 - Fix a sliding sync bug as a result of adding additional tests whereby the wrong room could be unsubscribed if `ViewRoom` calls occurred sufficiently quickly.

The next PR will focus on how to pass stores through to components in a non-global way.

## Checklist

* [x] Tests written for new code (and old code if feasible)
* [x] Linter and other CI checks pass
* [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

---
This change is marked as an *internal change* (Task), so will not be included in the changelog.

### CODE REVIEW

File: src/components/views/right_panel/TimelineCard.tsx [hunk 1]
```diff
@@ -92,7 +91,7 @@ export default class TimelineCard extends React.Component<IProps, IState> {
     }
 
     public componentDidMount(): void {
-        this.roomStoreToken = RoomViewStore.instance.addListener(this.onRoomViewStoreUpdate);
+        RoomViewStore.instance.addListener(UPDATE_EVENT, this.onRoomViewStoreUpdate);
```
Comment (dbkr) [inline src/components/views/right_panel/TimelineCard.tsx:94]: Minor: can we use on/addListener consistently? It's `on` above and I think that's our usual.

**Review by dbkr — COMMENTED** [review]
Looks like an improvement to me, but I'll let someone from the app team give the final stamp of approval.

File: src/components/views/right_panel/TimelineCard.tsx (hunk 1 above)
Comment (kegsay) [reply src/components/views/right_panel/TimelineCard.tsx:94]: Sure, I just kept with `addListener` because that's what it was using previously.

**Review by t3chguy — APPROVED** [review]
Looks sane

File: src/components/views/right_panel/TimelineCard.tsx (hunk 1 above)
Comment (kegsay) [reply src/components/views/right_panel/TimelineCard.tsx:94]: TimelineCard doesn't use `.on` anywhere afaict, so not going to change this.

**Pull request merged**